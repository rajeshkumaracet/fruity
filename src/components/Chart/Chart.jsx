import { useEffect, useRef } from "react";
import Chart from "chart.js/auto";

const PlotChart = ({ data }) => {
  const { incomeStatementData, balanceSheetData } = data;
  const chartRef = useRef(null);
  const chartInstance = useRef(null);

  useEffect(() => {
    if (
      !incomeStatementData ||
      incomeStatementData.length === 0 ||
      !balanceSheetData ||
      balanceSheetData.length === 0
    )
      return;

    const quarterlyNetIncome = incomeStatementData.map((entry) => ({
      quarter: entry.fiscalDateEnding.substring(0, 7),
      netIncome: parseFloat(entry.netIncome),
    }));

    const quarterlyTotalRevenue = incomeStatementData.map((entry) => ({
      quarter: entry.fiscalDateEnding.substring(0, 7),
      totalRevenue: parseFloat(entry.totalRevenue),
    }));

    const quarterlyShareholderEquity = balanceSheetData.map((entry) => ({
      quarter: entry.fiscalDateEnding.substring(0, 7),
      totalShareholderEquity: parseFloat(entry.totalShareholderEquity),
    }));

    const quarters = quarterlyNetIncome.map((entry) => entry.quarter);
    const netIncomeData = quarterlyNetIncome.map(
      (entry) => entry.netIncome || 0
    );
    const totalRevenueData = quarterlyTotalRevenue.map(
      (entry) => entry.totalRevenue || 0
    );
    const totalShareholderEquityData = quarterlyShareholderEquity.map(
      (entry) => entry.totalShareholderEquity || 0
    );

    if (chartInstance.current) {
      chartInstance.current.destroy(); // Destroy existing chart instance
    }

    const ctx = chartRef.current.getContext("2d");
    chartInstance.current = new Chart(ctx, {
      type: "line",
      data: {
        labels: quarters,
        datasets: [
          {
            label: "Quarterly Net Income",
            data: netIncomeData,
            borderColor: "rgba(255, 99, 132, 1)",
            backgroundColor: "rgba(255, 99, 132, 0.2)",
            fill: false,
          },
          {
            label: "Quarterly Total Revenue",
            data: totalRevenueData,
            borderColor: "rgba(54, 162, 235, 1)",
            backgroundColor: "rgba(54, 162, 235, 0.2)",
            fill: false,
          },
          {
            label: "Quarterly Total Shareholder Equity",
            data: totalShareholderEquityData,
            borderColor: "rgba(75, 192, 192, 1)",
            backgroundColor: "rgba(75, 192, 192, 0.2)",
            fill: false,
          },
        ],
      },
      options: {
        scales: {
          y: {
            beginAtZero: true,
          },
        },
      },
    });
  }, [balanceSheetData, incomeStatementData]);

  return <canvas ref={chartRef} />;
};

export default PlotChart;
