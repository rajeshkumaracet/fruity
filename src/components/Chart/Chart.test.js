import { render, screen } from "@testing-library/react";

import Chart from "./Chart";

describe("<Chart/>", () => {
  const props = {
    data: {
      incomeStatementData: [
        {
          fiscalDateEnding: "2022-01-01",
          netIncome: 100,
          totalRevenue: 200,
        },
        {
          fiscalDateEnding: "2022-04-01",
          netIncome: 150,
          totalRevenue: 250,
        },
      ],
      balanceSheetData: [
        { fiscalDateEnding: "2022-01-01", totalShareholderEquity: 300 },
        { fiscalDateEnding: "2022-04-01", totalShareholderEquity: 350 },
      ],
    },
  };

  it("should render correctly", () => {
    const { container } = render(<Chart {...props} />);
    expect(container).toMatchSnapshot();
  });

  it("renders a canvas element", () => {
    render(<Chart {...props} />);

    const canvasElement = screen.getByTestId("chart-canvas");
    expect(canvasElement).toBeInTheDocument();
  });
});
