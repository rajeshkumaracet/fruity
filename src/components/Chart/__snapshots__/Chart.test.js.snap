// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Chart/> should render correctly 1`] = `
<div>
  <canvas
    data-testid="chart-canvas"
  />
</div>
`;
